File: SymmetricTrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# All TC passed on leetocde

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:

        #Using BFS optimized approach
        #Time complexity - O(n) as we visit all elements of tree
        #Space complexity - O(n) - size of queue can reach max leaf nodes size i.e. n/2
        if not root:
            return True
        queue = collections.deque()
        queue.append(root.left)
        queue.append(root.right)

        while queue:
            left = queue.popleft()
            right = queue.popleft()

            if not left and not right:
                continue
            if not left or not right or left.val!=right.val:
                return False
            
            queue.append(left.left)
            queue.append(right.right)
            queue.append(left.right)
            queue.append(right.left)

        return True

    #----------------------------------OR-------------------------------------------------

        #Using BFS approach
        #Time complexity - O(n^2) as we check for each level if queue elements are palindrome
        #Space complexity - O(n) - size of queue can reach max leaf nodes size i.e. n/2
        def isPalindrome(arr):
            print(arr)
            i = 0
            j = len(arr)-1
            while i<=j:
                if not arr[i] and not arr[j]:
                    i+=1
                    j-=1
                    continue
                if not arr[i] or not arr[j]:
                    return False
                if arr[i].val!=arr[j].val:
                    return False
                i+=1
                j-=1
            return True

        if not root:
            return True
        queue = collections.deque()
        queue.append(root)
        while queue:
            if not isPalindrome(queue):
                    return False
            for i in range(len(queue)):
                cur = queue.popleft()
                if cur:
                    queue.append(cur.left)
                    queue.append(cur.right)
        return True

#----------------------------------OR-------------------------------------------------
       
        #Using DFS approach
        #Time complexity - O(n)
        #Space complexity - O(h)
        if not root:
            return True
        

        def helper(left, right):
            if not left and not right:
                return True
            
            if not left or not right or left.val!=right.val:
                return False

            l = helper(left.left, right.right)
            r = helper(left.right, right.left)

            return l and r

        return helper(root.left, root.right)

File: SymmetricTrees.java
// Time Complexity : O(n) no.of elememts in tree
// Space Complexity : O(n) for queue
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :
//Iterative
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root.left);
        q.add(root.right);

        while(!q.isEmpty()) {
            TreeNode left = q.poll();
            TreeNode right = q.poll();

            if(left == null && right == null) continue;
            if(left == null || right == null || left.val != right.val){
                return false;
            }

            q.add(left.left);
            q.add(right.right);
            q.add(left.right);
            q.add(right.left);
        }
        return true;
    }
}

// Time Complexity : O(N) n is # of elements in tree
// Space Complexity : O(h)  height of the tree for recurrsion.
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :
class Solution {
    boolean isValid;
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;
        isValid = true;

        dfs(root.left, root.right);

        return isValid;
    }

    private void dfs(TreeNode left, TreeNode right) {
        //base
        if(left == null && right == null) return;
        if(left == null || right == null || left.val != right.val){
            isValid = false;
            return;
        }
        //logic
        dfs(left.left, right.right);
        dfs(left.right, right.left);


    }
}